### PR TITLE
SPARKNLP-623: Refactoring loadSavedModel to a more generic/flexible method

### PR DESCRIPTION
**Pull request by maziyarpanahi**
This PR attempts to refactor loadSavedModel to make it more flexible to allow additional DL engines in the future. (This is a work-in-progress PR)

Tests:

- [x] Local test on Windows (Paths usually used on Windows)
- [x] Local test on Linux
- [x] Local test on macOS
- [x] local Google Colab
- [x] Databricks test (DBFS dbfs:/)
- [x] Databricks test (Driver file:/// or /)
- [x] Databricks test (load S3)
- [x] Local or Colab load from S3
- [x] GCP test (HDFS)

### CODE REVIEW

**Comment by DevinTDHa** [comment]
@maziyarpanahi I created a new draft PR here to compare the changes: https://github.com/JohnSnowLabs/spark-nlp/pull/12931

File: src/main/scala/com/johnsnowlabs/ml/util/LoadExternalModel.scala [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+package com.johnsnowlabs.ml.util
+
+import com.johnsnowlabs.nlp.util.io.{ExternalResource, ReadAs, ResourceHelper}
+
+import java.io.File
+
+object LoadExternalModel {
+
+  def modelSanityCheck(modelPath: String): String = {
```
Comment (DevinTDHa) [inline src/main/scala/com/johnsnowlabs/ml/util/LoadExternalModel.scala:25]: I moved the check for vocab.txt here as well (and potentially other assets). Is this something that we would want?

https://github.com/JohnSnowLabs/spark-nlp/pull/12931/files#diff-13282f334c1c1e8d9a1253a31cce7d87da16659719b1b606f251e6688e420a75R23

**Comment by maziyarpanahi** [comment]
> @maziyarpanahi I created a new draft PR here to compare the changes: #12931

I am going to finish it today and then we can have a look at the overall picture. (I will do the same for sentencepiece, labels, merges, etc.)

File: src/main/scala/com/johnsnowlabs/nlp/embeddings/BertEmbeddings.scala [hunk 2]
```diff
@@ -402,46 +401,41 @@ trait ReadBertTensorflowModel extends ReadTensorflowModel {
 
   addReader(readTensorflow)
 
-  def loadSavedModel(tfModelPath: String, spark: SparkSession): BertEmbeddings = {
+  def loadSavedModel(modelPath: String, spark: SparkSession): BertEmbeddings = {
 
-    val f = new File(tfModelPath)
-    val savedModel = new File(tfModelPath, "saved_model.pb")
+    val detectedEngine = modelSanityCheck(modelPath)
```
Comment (DevinTDHa) [inline src/main/scala/com/johnsnowlabs/nlp/embeddings/BertEmbeddings.scala:406]: We will have to merge this with the remote loading of the model:

https://github.com/JohnSnowLabs/spark-nlp/pull/12931/files#diff-13282f334c1c1e8d9a1253a31cce7d87da16659719b1b606f251e6688e420a75R39

**Review by DevinTDHa — COMMENTED** [review]
Should we merge it for BertEmbeddings first, before continuing to implement it for the rest of the annotators?

File: src/main/scala/com/johnsnowlabs/nlp/embeddings/BertEmbeddings.scala (hunk 2 above)
Comment (maziyarpanahi) [reply src/main/scala/com/johnsnowlabs/nlp/embeddings/BertEmbeddings.scala:406]: @DevinTDHa We will. I am refactoring all of the loadSavedModel first, then I will see how your DBFS/S3 implementation.

I suggest you don't refactor, make one of them work/compatible with DBFS/S3/HDFS, and once you are done testing I can see the best place to add those changes

**Comment by maziyarpanahi** [comment]
> Should we merge it for BertEmbeddings first, before continuing to implement it for the rest of the annotators?

I made this PR so you won't have to refactor, I say you make sure the DBFS/S3/HDFS work (let's say you pick one annotator like BertEmbeddings), and I continue with the refactoring here. Once we are both done I will merge your changes into this branch.

**Pull request merged**